File: arrays/two-sum.java
// 1. Two Sum
// https://leetcode.com/problems/two-sum/

import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (map.get(target - nums[i]) != null)
                return new int[] { i, map.get(target - nums[i]) };
            map.put(nums[i], i);
        }
        return new int[] { -1, -1 };
    }
}

// Better Solution
// Init a map -> key (number) : value (index)
// Iterate through the list
// we check if (target - current) is in map
// if in map, we have our 2 indices already
// if not in map, add the number:index into the map
// O(n)

File: two_sum.js
// 1. Two Sum
// https://leetcode.com/problems/two-sum/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  const map = {};

  for (let i = 0; i < nums.length; i++) {
    if (target - nums[i] in map) return [i, map[target - nums[i]]];
    map[nums[i]] = i;
  }
};

// Time: O(n)
// Space: O(n)
// init map where k:v is the element:index
// Iterate through the array
// Check in the map if (target - element) exists in the map
// if exists, we have found the pair
// if doesnt exist, Store each element in a map
